storage-proofs/porep/src/stacked/vanilla/proof.rs

```diff
@@ -451,39 +462,41 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
                             chunked_nodes_count
                         ];
 
-                    rayon::scope(|s| {
-                        let n = num_cpus::get();
-
-                        // only split if we have at least two elements per thread
-                        let num_chunks = if n > chunked_nodes_count * 2 { 1 } else { n };
+                    // Allocate layer data array and insert a placeholder for each layer.
+                    let mut layer_data = Vec::with_capacity(layers);
+                    for _ in 0..layers {
+                        layer_data.push(Vec::new());
+                    }
 
-                        // chunk into n chunks
-                        let chunk_size =
-                            (chunked_nodes_count as f64 / num_chunks as f64).ceil() as usize;
+                    rayon::scope(|s| {
+                        // capture a shadowed version of layer_data.
+                        let layer_data: &mut Vec<_> = &mut layer_data;
 
-                        // gather all n chunks in parallel
-                        for (chunk, columns_chunk) in columns.chunks_mut(chunk_size).enumerate() {
+                        // gather all layer data in parallel.
+                        s.spawn(move |_| {
                             let labels = &labels;
+                            for k in 1..=layers {
+                                let store = labels.labels_for_layer(k);
+                                let start = (i * nodes_count) + node_index;
+                                let end = start + chunked_nodes_count;
+                                let elements: Vec<<Tree::Hasher as Hasher>::Domain> =
+                                    store.read_range(std::ops::Range { start, end }).unwrap();
+
+                                let fr_elements: Vec<_> =
+                                    elements.into_iter().map(|el| el.into()).collect();
+                                layer_data[k - 1] = fr_elements;
+                            }
+                        });
+                    });
```

Comment:
could be 

```rust
for layer_elements in layer_data.iter_mut() {
    // ... 
    layer_elements.extend(elements.into_iter().map(Into::into));
}
```

which should avoid some allocations

Reply:
I'll take a look at this, thanks

Reply:
I see what you're saying now (along with the above comment).  Thanks!